### ISSUE TITLE
Proposal: add 'allowOnTop' option to no-param-reassign rule

### ISSUE THREAD

**Issue by farin** [eslint/eslint#3652]
When allowOnTop is true, reassign is allowed at the beginning of function scope (same concept like var statement in vars-on-top rule).

The reason - I don't want to reassign params except intended setting default like:
```js
function (data) {  
   data = data || 0;
   ...
}
```
This can be avoided in ES6 with default arguments, but they can't used in all cases, for instance making a defensive copy:
```js
function (data) {  
   data = _.clone(data);
   ...
}
```
You can use `_` prefix or new variable `dataCopy` but i don't like it much. In fact I want to hide original variable to prevent unwanted access to unsafe origin.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by ilyavolodin** [comment]
What should happen in the case like this:
```js
function (data, options) {
  var data = data;
  if (options) {
     var options = { items: options, data: data };
  }
}
```
Should options override be flagged?

**Comment by farin** [comment]
@ilyavolodin 
do you mean `items: options.items`
in such case I am using

```
function (data, options) {
   options = Object.assign({ data: data }, options);
   ...
}
```
so I still ok with allow just pure top and flag your example.

**Comment by zallek** [comment]
:+1: I'd really like to have this.

**Comment by gyandeeps** [comment]
@eslint/eslint-team thoughts?

**Comment by nzakas** [comment]
On the fence. Anyone want to champion this?

**Comment by strburst** [comment]
:+1: Anecdotally, I ran into this exact issue just now, so I'd also love to have this.

(Incidentally, thanks @farin for your blog post on extending AirBnB's style guide; I found it helpful.)

**Comment by nzakas** [comment]
We still need a champion from the ESLint team for this enhancement.

**Comment by nzakas** [comment]
We were unable to find a champion on the team for this proposal, so closing.

**Comment by abecks** [comment]
Do these issues get re-opened at any point? Or are they lost to their watery grave because no one was available at the time to implement?

**Comment by ilyavolodin** [comment]
@abecks It's not about the time. It's about nobody on the team feeling strongly enough that this exception is important and widespread to champion or support it. That doesn't mean that you couldn't implement it yourself as a plugin however. Not everything makes it into the core, but that's exactly why ESLint was designed from the start to support plugins.

**Comment by ljharb** [comment]
For future explorers: v8 [severely deoptimizes a function in sloppy mode](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#31-reassigning-a-defined-parameter-while-also-mentioning-arguments-in-the-body-in-sloppy-mode-only-typical-example) when you assign arguments, even on top, so it's potentially a big performance hazard - not "just" style.

**Comment by romellem** [comment]
@ljharb That is only if you reference the `arguments` variable in the body of your function.

From the link you gave:

> Reassigning a defined parameter **while also mentioning `arguments` in the body** [causes the function to be unoptimizable.]

Some example files:

---

*unoptimized.js*  
```javascript
function test(a) {
    if (arguments.length < 1) a = 10;
}

console.time('test');

for (let i = 0; i < 100000000; i++) {
    test();
}

console.timeEnd('test');

// prints "test: 3000ms"
```

---

*optimized.js*  
```javascript
function test(a) {
    a = a || 10;
}

console.time('test');

for (let i = 0; i < 100000000; i++) {
    test();
}

console.timeEnd('test');

// prints "test: 230ms"
```

---

As you can see, the file *optimized.js* with the function that has the `a = a || 10;` line runs about **13 times** faster, and is in fact being optimized by V8.

**Comment by ljharb** [comment]
@romellem fair enough; but variables are free. there's still no good reason to reassign to named arguments. Your test lacks an example that does `var b = a || 10;` instead.

**Comment by cipri-tom** [comment]
would it be feasible to have `no-param-reassign` recognise that some params are literals so assigning or changing them would not have side effects ?

```javascript
function fromBytes(dView, startOffset = 0) {
    while (startOffset < dView.byteLength) {
        // magic bytes extraction with dView.getUint8(startOffset)
        startOffset += 1;  // offending line :(
    }
}
```

**Comment by ljharb** [comment]
@cipri-tom reassignment doesn't have side effects; primitives are immutable (so "changing them" doesn't apply). Reassignment isn't bad because it has side effects; it's bad because you should just make a new variable, for clarity.
